### PR TITLE
Linkis1.2.0[datamodel][dataassets][datawarehouse]

### PR DESCRIPTION
**Pull request by Zosimer**
Linkis1.2.0+dss1.2.0[datamodel][dataassets][datawarehouse],Realize data model management, connect with Atlas, and realize blood relationship management

### What is the purpose of the change
(For example: AppConn Core defines the abstractions and interfaces of the AppConn core functions.
It is the core concept that enables DSS to easily and quickly integrate various upper-layer web systems.
Related issues: #953. )

### CODE REVIEW

File: dss-apps/dss-datamodel-center/dss-datamodel-dimension-management/src/main/java/com/webank/wedatasphere/dss/datamodel/dimension/service/impl/DimensionServiceImpl.java [hunk 1]
```diff
@@ -0,0 +1,214 @@
...
+                LOGGER.error("errorCode : {}, dimension field identifier can not repeat", ErrorCode.DIMENSION_UPDATE_ERROR.getCode());
+                throw new DSSDatamodelCenterException(ErrorCode.DIMENSION_UPDATE_ERROR.getCode(), "dimension field identifier can not repeat ");
+            }
+        }
+        DssDatamodelDimension updateOne = modelMapper.map(vo,DssDatamodelDimension.class);
+        updateOne.setUpdateTime(new Date());
+        getBaseMapper().update(updateOne, Wrappers.<DssDatamodelDimension>lambdaUpdate().eq(DssDatamodelDimension::getId,id));
+
+        //同步atlas
+        UpdateModelTypeResult updateModelTypeResult = assertsSyncService.syncUpdateModel(
```
Comment (Dlimeng) [inline dss-apps/dss-datamodel-center/dss-datamodel-dimension-management/src/main/java/com/webank/wedatasphere/dss/datamodel/dimension/service/impl/DimensionServiceImpl.java:138]: publisher.publishEvent(new UpdateModelEvent(this
                ,DataModelSecurityContextHolder.getContext().getDataModelAuthentication().getUser()
                ,vo.getFieldIdentifier()
                ,orgFieldIdentifier
                ,ClassificationConstant.DIMENSION));

File: dss-apps/dss-datamodel-center/dss-datamodel-dimension-management/src/main/java/com/webank/wedatasphere/dss/datamodel/dimension/service/impl/DimensionServiceImpl.java [hunk 2]
```diff
@@ -0,0 +1,214 @@
...
+        if (dssDatamodelDimension == null){
+            throw new DSSDatamodelCenterException(ErrorCode.DIMENSION_DELETE_ERROR.getCode(), "dimension id " +id +" not exists");
+        }
+        //校验引用情况
+        if(datamodelReferencService.dimensionReferenceCount(dssDatamodelDimension.getName())>0||datamodelReferencService.dimensionReferenceCount(dssDatamodelDimension.getFieldIdentifier())>0){
+              throw new DSSDatamodelCenterException(ErrorCode.DIMENSION_DELETE_ERROR.getCode(), "dimension id " +id +" has referenced");
+        }
+        getBaseMapper().deleteById(id);
+        //同步资产
+        assertsSyncService.syncDeleteModel(new DeleteModelEvent(this
```
Comment (Dlimeng) [inline dss-apps/dss-datamodel-center/dss-datamodel-dimension-management/src/main/java/com/webank/wedatasphere/dss/datamodel/dimension/service/impl/DimensionServiceImpl.java:162]: publisher.publishEvent(new DeleteModelEvent(this
                ,DataModelSecurityContextHolder.getContext().getDataModelAuthentication().getUser()
                ,dssDatamodelDimension.getFieldIdentifier()
                ,ClassificationConstant.DIMENSION));

File: dss-apps/dss-datamodel-center/dss-datamodel-indicator-management/src/main/java/com/webank/wedatasphere/dss/datamodel/indicator/service/impl/IndicatorServiceImpl.java [hunk 3]
```diff
@@ -0,0 +1,539 @@
...
+        newOne.setUpdateTime(new Date());
+        if (!save(newOne)) {
+            LOGGER.error("errorCode : {}, add indicator error", ErrorCode.INDICATOR_ADD_ERROR.getCode());
+            throw new DSSDatamodelCenterException(ErrorCode.INDICATOR_ADD_ERROR.getCode(), "add indicator error");
+        }
+
+        indicatorContentService.addIndicatorContent(newOne.getId(), version, vo.getContent());
+
+        //同步atlas
+        CreateModelTypeResult result = assertsSyncService.syncCreateModel(
```
Comment (Dlimeng) [inline dss-apps/dss-datamodel-center/dss-datamodel-indicator-management/src/main/java/com/webank/wedatasphere/dss/datamodel/indicator/service/impl/IndicatorServiceImpl.java:113]: publisher.publishEvent(  new CreateModelEvent(this
                , DataModelSecurityContextHolder.getContext().getDataModelAuthentication().getUser()
                , vo.getFieldIdentifier()
                , ClassificationConstant.INDICATOR));

File: dss-apps/dss-datamodel-center/dss-datamodel-indicator-management/src/main/java/com/webank/wedatasphere/dss/datamodel/indicator/service/impl/IndicatorServiceImpl.java [hunk 4]
```diff
@@ -0,0 +1,539 @@
...
+            LOGGER.error("errorCode : {}, update indicator error not exists", ErrorCode.INDICATOR_UPDATE_ERROR.getCode());
+            throw new DSSDatamodelCenterException(ErrorCode.INDICATOR_UPDATE_ERROR.getCode(), "update indicator error not exists");
+        }
+
+        indicatorContentService.updateIndicatorContent(orgId, version, vo.getContent());
+
+
+
+        //同步atlas
+        UpdateModelTypeResult updateModelTypeResult = assertsSyncService.syncUpdateModel(
```
Comment (Dlimeng) [inline dss-apps/dss-datamodel-center/dss-datamodel-indicator-management/src/main/java/com/webank/wedatasphere/dss/datamodel/indicator/service/impl/IndicatorServiceImpl.java:182]: publisher.publishEvent(new UpdateModelEvent(this
                ,DataModelSecurityContextHolder.getContext().getDataModelAuthentication().getUser()
                ,vo.getFieldIdentifier()
                ,orgFieldIdentifier
                ,ClassificationConstant.INDICATOR));

File: dss-apps/dss-datamodel-center/dss-datamodel-indicator-management/src/main/java/com/webank/wedatasphere/dss/datamodel/indicator/service/impl/IndicatorServiceImpl.java [hunk 5]
```diff
@@ -0,0 +1,539 @@
...
+        //同时删除版本
+        indicatorVersionService.getBaseMapper()
+                .delete(Wrappers.<DssDatamodelIndicatorVersion>lambdaQuery().eq(DssDatamodelIndicatorVersion::getName, dssDatamodelIndicator.getName()));
+        //删除指标
+        getBaseMapper().deleteById(id);
+        indicatorContentService.getBaseMapper().delete(Wrappers.<DssDatamodelIndicatorContent>lambdaQuery().eq(DssDatamodelIndicatorContent::getIndicatorId, id));
+
+
+        //同步资产
+        assertsSyncService.syncDeleteModel(new DeleteModelEvent(this
```
Comment (Dlimeng) [inline dss-apps/dss-datamodel-center/dss-datamodel-indicator-management/src/main/java/com/webank/wedatasphere/dss/datamodel/indicator/service/impl/IndicatorServiceImpl.java:222]: publisher.publishEvent(new DeleteModelEvent(this
                ,DataModelSecurityContextHolder.getContext().getDataModelAuthentication().getUser()
                ,dssDatamodelIndicator.getFieldIdentifier()
                ,ClassificationConstant.INDICATOR));

File: dss-apps/dss-datamodel-center/dss-datamodel-measure-management/src/main/java/com/webank/wedatasphere/dss/datamodel/measure/service/impl/MeasureServiceImpl.java [hunk 6]
```diff
@@ -0,0 +1,214 @@
...
+            LOGGER.error("errorCode : {}, measure name can not be referenced ", ErrorCode.MEASURE_UPDATE_ERROR.getCode());
+            throw new DSSDatamodelCenterException(ErrorCode.MEASURE_UPDATE_ERROR.getCode(), "measure name can not be referenced");
+        }
+
+        DssDatamodelMeasure newOne = modelMapper.map(vo, DssDatamodelMeasure.class);
+        newOne.setCreateTime(new Date());
+        newOne.setUpdateTime(new Date());
+        getBaseMapper().insert(newOne);
+        //同步atlas
+        CreateModelTypeResult result = assertsSyncService.syncCreateModel(
```
Comment (Dlimeng) [inline dss-apps/dss-datamodel-center/dss-datamodel-measure-management/src/main/java/com/webank/wedatasphere/dss/datamodel/measure/service/impl/MeasureServiceImpl.java:78]: publisher.publishEvent( new CreateModelEvent(this
                , DataModelSecurityContextHolder.getContext().getDataModelAuthentication().getUser()
                , vo.getFieldIdentifier()
                , ClassificationConstant.MEASURE));

File: dss-apps/dss-datamodel-center/dss-datamodel-measure-management/src/main/java/com/webank/wedatasphere/dss/datamodel/measure/service/impl/MeasureServiceImpl.java [hunk 7]
```diff
@@ -0,0 +1,214 @@
...
+                throw new DSSDatamodelCenterException(ErrorCode.MEASURE_UPDATE_ERROR.getCode(), "measure field identifier can not repeat or referenced");
+            }
+        }
+        DssDatamodelMeasure updateOne = modelMapper.map(vo, DssDatamodelMeasure.class);
+        updateOne.setUpdateTime(new Date());
+        getBaseMapper().update(updateOne, Wrappers.<DssDatamodelMeasure>lambdaUpdate().eq(DssDatamodelMeasure::getId, id));
+        //同步atlas
+        UpdateModelTypeResult updateModelTypeResult = assertsSyncService.syncUpdateModel(
+                new UpdateModelEvent(this
+                        , DataModelSecurityContextHolder.getContext().getDataModelAuthentication().getUser()
```
Comment (Dlimeng) [inline dss-apps/dss-datamodel-center/dss-datamodel-measure-management/src/main/java/com/webank/wedatasphere/dss/datamodel/measure/service/impl/MeasureServiceImpl.java:140]: publisher.publishEvent( new UpdateModelEvent(this
                , DataModelSecurityContextHolder.getContext().getDataModelAuthentication().getUser()
                , vo.getFieldIdentifier()
                , orgFieldIdentifier
                , ClassificationConstant.MEASURE));

File: dss-apps/dss-datamodel-center/dss-datamodel-measure-management/src/main/java/com/webank/wedatasphere/dss/datamodel/measure/service/impl/MeasureServiceImpl.java [hunk 8]
```diff
@@ -0,0 +1,214 @@
...
+        }
+        //校验引用情况
+        if(datamodelReferencService.measureReferenceCount(dssDatamodelMeasure.getName())>0
+                ||datamodelReferencService.measureReferenceCount(dssDatamodelMeasure.getFieldIdentifier())>0){
+            throw new DSSDatamodelCenterException(ErrorCode.MEASURE_DELETE_ERROR.getCode(), "measure id " + id + " has referenced");
+        }
+
+        getBaseMapper().deleteById(id);
+        //同步资产
+        assertsSyncService.syncDeleteModel(new DeleteModelEvent(this
```
Comment (Dlimeng) [inline dss-apps/dss-datamodel-center/dss-datamodel-measure-management/src/main/java/com/webank/wedatasphere/dss/datamodel/measure/service/impl/MeasureServiceImpl.java:163]: publisher.publishEvent(new DeleteModelEvent(this
                , DataModelSecurityContextHolder.getContext().getDataModelAuthentication().getUser()
                , dssDatamodelMeasure.getFieldIdentifier()
                , ClassificationConstant.MEASURE));

File: dss-apps/dss-datamodel-center/dss-datamodel-table-management/src/main/java/com/webank/wedatasphere/dss/datamodel/table/service/impl/LabelServiceImpl.java [hunk 9]
```diff
@@ -0,0 +1,203 @@
...
+
+        DssDatamodelLabel newOne = modelMapper.map(vo,DssDatamodelLabel.class);
+        newOne.setParams(gson.toJson(vo.getParamMap()));
+        newOne.setCreateTime(new Date());
+        newOne.setUpdateTime(new Date());
+
+        getBaseMapper().insert(newOne);
+
+        //todo 同步资产创建标签
+        assertsSyncService.syncCreateLabel(new CreateLabelEvent(this
```
Comment (Dlimeng) [inline dss-apps/dss-datamodel-center/dss-datamodel-table-management/src/main/java/com/webank/wedatasphere/dss/datamodel/table/service/impl/LabelServiceImpl.java:78]: publisher.publishEvent(new CreateLabelEvent(this
                , DataModelSecurityContextHolder.getContext().getDataModelAuthentication().getUser()
                , vo.getName()));

File: dss-apps/dss-datawarehouse-design/dss-data-warehouse-service/src/main/java/com/webank/wedatasphere/warehouse/service/impl/DwLayerServiceImpl.java [hunk 10]
```diff
@@ -0,0 +1,523 @@
...
+        layer.setCreateTime(now);
+        layer.setUpdateTime(now);
+        layer.setStatus(Boolean.TRUE);
+        layer.setLockVersion(1L);
+
+        int insert = this.dwLayerMapper.insert(layer);
+        PreconditionUtil.checkState(1 == insert, DwException.stateReject("create dw layer failed"));
+
+        // 建立关联
+        try {
```
Comment (Dlimeng) [inline dss-apps/dss-datawarehouse-design/dss-data-warehouse-service/src/main/java/com/webank/wedatasphere/warehouse/service/impl/DwLayerServiceImpl.java:145]: Asynchronous association

File: dss-apps/dss-datawarehouse-design/dss-data-warehouse-service/src/main/java/com/webank/wedatasphere/warehouse/service/impl/DwLayerServiceImpl.java [hunk 11]
```diff
@@ -0,0 +1,523 @@
...
+//            return Message.ok();
+//        }
+//        layer.setStatus(Boolean.FALSE);
+//        int i = this.dwLayerMapper.updateById(layer);
+        int i = this.dwLayerMapper.deleteById(layer);
+        PreconditionUtil.checkState(1 == i, DwException.stateReject("remove action failed"));
+
+        // 删除关联
+        try {
+            LinkisDataAssetsRemoteClient dataAssetsRemoteClient = LinkisRemoteClientHolder.getDataAssetsRemoteClient();
```
Comment (Dlimeng) [inline dss-apps/dss-datawarehouse-design/dss-data-warehouse-service/src/main/java/com/webank/wedatasphere/warehouse/service/impl/DwLayerServiceImpl.java:342]: Asynchronous association

File: dss-apps/dss-datawarehouse-design/dss-data-warehouse-service/src/main/java/com/webank/wedatasphere/warehouse/service/impl/DwLayerServiceImpl.java [hunk 12]
```diff
@@ -0,0 +1,523 @@
...
+        UpdateWrapper<DwLayer> updateWrapper = new UpdateWrapper<>();
+        updateWrapper.eq("id", layer.getId());
+        updateWrapper.eq("lock_version", oldVersion);
+
+        int i = this.dwLayerMapper.update(layer, updateWrapper);
+
+        PreconditionUtil.checkState(1 == i, DwException.stateReject("update layer failed"));
+
+        // 更新关联
+        if (!Objects.equals(orgEnName, layer.getEnName())) {
```
Comment (Dlimeng) [inline dss-apps/dss-datawarehouse-design/dss-data-warehouse-service/src/main/java/com/webank/wedatasphere/warehouse/service/impl/DwLayerServiceImpl.java:465]: Asynchronous association

File: dss-apps/dss-datawarehouse-design/dss-data-warehouse-service/src/main/java/com/webank/wedatasphere/warehouse/service/impl/DwThemeDomainServiceImpl.java [hunk 13]
```diff
@@ -0,0 +1,431 @@
...
+        record.setUpdateTime(now);
+        record.setStatus(Boolean.TRUE);
+        record.setLockVersion(1L);
+
+        int insert = this.dwThemeDomainMapper.insert(record);
+        PreconditionUtil.checkState(1 == insert, DwException.stateReject("create dw theme domain failed"));
+
+        // 建立关联
+        try {
+            LinkisDataAssetsRemoteClient dataAssetsRemoteClient = LinkisRemoteClientHolder.getDataAssetsRemoteClient();
```
Comment (Dlimeng) [inline dss-apps/dss-datawarehouse-design/dss-data-warehouse-service/src/main/java/com/webank/wedatasphere/warehouse/service/impl/DwThemeDomainServiceImpl.java:211]: Asynchronous association

File: dss-apps/dss-datawarehouse-design/dss-data-warehouse-service/src/main/java/com/webank/wedatasphere/warehouse/service/impl/DwThemeDomainServiceImpl.java [hunk 14]
```diff
@@ -0,0 +1,431 @@
...
+//        if (Objects.equals(Boolean.FALSE, record.getStatus())) {
+//            return Message.ok();
+//        }
+//        record.setStatus(Boolean.FALSE);
+//        int i = this.dwThemeDomainMapper.updateById(record);
+        // physical delete, because of name unique
+        int i = this.dwThemeDomainMapper.deleteById(record);
+        PreconditionUtil.checkState(1 == i, DwException.stateReject("remove action failed"));
+
+        // 删除关联
```
Comment (Dlimeng) [inline dss-apps/dss-datawarehouse-design/dss-data-warehouse-service/src/main/java/com/webank/wedatasphere/warehouse/service/impl/DwThemeDomainServiceImpl.java:270]: Asynchronous association

**Review by Dlimeng — COMMENTED** [review]
Many of them are the source files of my original version. Check a few key points and suggest modifications.

File: dss-apps/dss-datamodel-center/dss-datamodel-table-management/src/main/java/com/webank/wedatasphere/dss/datamodel/table/service/impl/TableServiceImpl.java [hunk 15]
```diff
@@ -0,0 +1,836 @@
...
+    private int referenceCount(String name, ModeType modeType) {
+        int currentCount = tableColumnsService.modelReferenceCount(modeType, name);
+        int currentCountEn = tableColumnsService.modelReferenceCountEn(modeType, name);
+//        int versionCount = tableVersionService.tableColumnsReference(name);
+
+        List<DssDatamodelTableVersion> preReferences = tableVersionService.tableColumnsReference(name);
+        int versionCount = (int) preReferences.stream().filter(e->{
+            DssDatamodelTableColumns temp = gson.fromJson(e.getTableParams(),DssDatamodelTableColumns.class);
+            //名称相同且类型相同
+            return modeType.getType() == temp.getModelType()&&StringUtils.equals(name,temp.getModelName());
```
Comment (Dlimeng) [inline dss-apps/dss-datamodel-center/dss-datamodel-table-management/src/main/java/com/webank/wedatasphere/dss/datamodel/table/service/impl/TableServiceImpl.java:733]: temp.getModelType() !=null && modeType.getType() == temp.getModelType() && StringUtils.equals(name, temp.getModelName());

**Review by zqburde — APPROVED** [review]
LGTM

**Pull request merged**